### PR TITLE
New --emitExtension and --noImplicitExtensionName compiler options

### LINKED ISSUE #16577: Provide a way to add the '.js' file extension to the end of module specifiers
In order to use es6 modules in the browser, you need a .js file extension. However output doesn't add it.

In ts:
`import { ModalBackground } from './ModalBackground';`
In ES2015 output:
`import { ModalBackground } from './ModalBackground';`

Ideally I would like this to be output
`import { ModalBackground } from './ModalBackground.js';`

That way I can use the output in Chrome 51

```js
<!DOCTYPE html>
<html>
<head>
  <meta charset="UTF-8">
  <title>Webpack boilerplate</title>
  <script type="module" src="index.js"></script>
</head>
<body></body>
</html>
```
image

Related to https://github.com/Microsoft/TypeScript/issues/13422

### LINKED ISSUE #18442: Support `.mjs` output
Experimental support for ES modules just landed in Node 8.5 ([changelog](https://github.com/nodejs/node/blob/master/doc/changelogs/CHANGELOG_V8.md#8.5.0), [PR](https://github.com/nodejs/node/pull/14369)).
Since ES modules have some parsing and semantic differences, Node decided to use the `mjs` extension for ES modules (while `js` is for the "script" target and commonjs modules).

The current Typescript version (`2.5.2`) supports ES modules emission but uses the `js` extension by default. It means that to use it with Node, a post-compilation step is required to change the extension from `js` to `mjs`. This adds undesirable complexity to use native ES modules support with Node.

A solution would be to add a compiler option to output `*.mjs` files when emitting  ES modules.

**Edit (2018-03-22)**: The propositions below are a bit outdated. I recommend reading the issue to see the progression. See [this comment](https://github.com/Microsoft/TypeScript/issues/18442#issuecomment-365691593) for my current proposition.

Notes:
- It is possible that new parse goals will be added in the next versions of the ES spec. Letting the user provide the extension explicitly (as opposed to a boolean switch) would allow to be forward compatible with other new extension.
- It should still be possible to emit ES modules with the `*.js` extension, many tools rely on the `js` extension.
- Another solution would be to have a different extension for the source files: `*.mts` files would compile to `*.mjs`, this would be similar to `*.tsx` and `*.jsx`.

### LINKED ISSUE #30076: Proposal: emit files without `.ts` extension to get custom file extension
## Search Terms

- file extension
- .mjs
- .esm

## Suggestion

It is very useful if you can customize the output extension.
But mapping extension is boring and painful.
Maybe there is another way to make this easier.
If we don't map the extension, we just erase the `.ts` or `.tsx` extension to achieve it.
A compilerOptions is needed to prevent break existing TypeScript code.

## Use Cases

- Output the `.mjs` file for Node Module,
- Output `.es` or other file extensions without additional steps. (e.g. useful for tsc watching mode)

## Examples

input:

```ts
// src/lib.js.ts
export const result = 42;
```

```ts
// src/index.js.ts
import { result } from "./lib.js";

console.log(result);
```

output:

```ts
// build/lib.js
export const result = 42;
```

```ts
// build/index.js
import { result } from "./lib.js";

console.log(result);
```

---

input:

```ts
// src/lib.mjs.ts
export const result = 42;
```

```ts
// src/index.mjs.ts
import { result } from "./lib.mjs";

console.log(result);
```

output:

```ts
// build/lib.mjs
export const result = 42;
```

```ts
// build/index.mjs
import { result } from "./lib.mjs";

console.log(result);
```

---

input:

```ts
// src/lib.myext.ts
export const result = 42;
```

```ts
// src/index.myext.ts
import { result } from "./lib.myext";

console.log(result);
```

output:

```ts
// build/lib.myext
export const result = 42;
```

```ts
// build/index.myext
import { result } from "./lib.myext";

console.log(result);
```

## Checklist

My suggestion meets these guidelines:

- [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
- [x] This wouldn't change the runtime behavior of existing JavaScript code
- [x] This could be implemented without emitting different JS based on the types of the expressions
- [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
- [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

### LINKED ISSUE #33306: TypeScript ES modules incompatible with browser and Node.js
Typescript is incompatible with ECMAScript modules, as implemented in browsers and Node.js (the two most common ES runtimes).

**TypeScript Version:**  3.6.2 (Node.js 12.7.0)

**Search Terms:** ES2015 module, browser, Node.js

**Code**

https://github.com/pauldraper/ts-es-import

```sh
yarn install
yarn run build
```

A.

```sh
yarn run run
```

B.

```sh
yarn run serve
```

Browse to http://localhost:8080 with ES2015-compatible browser (Chrome, Edge, Firefox, Safari).

**Expected behavior:**

A.

"Hello world" is printed in the console.

B.

"Hello world" is printed.

**Actual behavior:**

A.

> GET http://localhost:8080/out/lib net::ERR_ABORTED 404 (Not Found)

B.

Error: Cannot find module '/home/paul/dev/pauldraper/typescript-nodejs/out/lib' imported from /home/paul/dev/pauldraper/typescript-nodejs/out/main.js

**Playground Link:**

**Related Issues:**

---

The issue is that while Node's CommonJS modules use a list of implicit file extensions, Node and browsers require the extension.

And neither Typescript's classic nor node resolution strategies permit a file extension. This choice allows tsc to not touch the import specifiers, as the .ts/.d.ts/.tsx/.js/.jsx/.json relationship stays the same before and after compilation.

**Solutions:**

1. Nothing. End-user maintains a post-process that either modifies the outputted import specifiers, or modifies the outputted file locations.

2. TypeScript adds a new --module value ESBrowser, which transforms the output import specifiers to have .js extension.

3. TypeScript adds a new option --outputExtension which when false or 'remove', strips the extension from the outputted file.

### PR DESCRIPTION
**Pull request by Jack-Works**
Related: #8529, #27957, #35749

fixes #30076, fixes #18442, fixes #16577, fixes #33306

See also: Node.js new ESModule support: https://medium.com/@nodejs/announcing-core-node-js-support-for-ecmascript-modules-c5d6dc29b663

# Summary

image

Improve support for `deno` and `browser` style import. Note: This pr doesn't include URI style import resolution (`https://deno.land/...` or `std:kv-storage`...), about URI style import, see  https://github.com/microsoft/TypeScript/issues/35749

## New `--noImplicitExtensionName` compiler options.

With this option on, the following code becomes invalid.

This new options will enforce developer to write valid import path on `deno` (`noImplicitExtensionName + emitExtension=.ts`) and ES Module for browser (`noImplicitExtensionName + emitExtension=.js (default value)`).

## New `--emitExtension` compiler options.
This flag allows tsc to emit `.mjs`, `.es.js`, `.es`, `.ts` or any extension user specifies.

tsconfig.json
```jsonc
{ compilerOptions: { emitExtension: ".mjs", noImplicitExtensionName: true } }
```

Source files:

Emitted files:

## Discussions
According to https://github.com/microsoft/TypeScript/issues/35163#issuecomment-558918600 maybe `--noImplicitExtensionName` should be default to `true` when the target module system is ESModule

# Details
Due to the ts meeting note for this pr (https://github.com/microsoft/TypeScript/issues/35589 )
I decided to rewrite ts pr to follow the ts design principle.

- [x] Resolve TS file by `--emitExtension`
- [x] Type error for import with no extension name
- [x] Emit file as ext of `emitExtension`
- [x] Allow import `.ts` and `.tsx` file when `emitExtension` === `.ts` or `.tsx`
- [x] If the extension name !== emitExtension, emit a compile error

### Resolve TS file by `--emitExtension`

Resolve type of `import './x.mjs'` to the file `x.ts`.
Will NOT rewrite the emitted file path.

### CODE REVIEW

**Comment by Jack-Works** [comment]
help wanted! how can I accept new baseline for this test?

```
       
         fourslash tests
           tests/cases/fourslash/completionListIsGlobalCompletion.ts
             fourslash test completionListIsGlobalCompletion.ts runs correctly:
     Error: At 1: Expected no completions
      at TestState.raiseError (src/harness/fourslash.ts:489:19)
      at TestState.verifyCompletionsWorker (src/harness/fourslash.ts:808:61)
      at TestState.verifyCompletions (src/harness/fourslash.ts:770:26)
      at Verify.completions (src/harness/fourslash.ts:3976:28)
      at eval (eval at runCode (src/harness/fourslash.ts:3290:23), <anonymous>:37:8)
      at runCode (src/harness/fourslash.ts:3291:13)
      at runFourSlashTestContent (src/harness/fourslash.ts:3271:9)
      at Object.runFourSlashTest (src/harness/fourslash.ts:3256:9)
      at Context.<anonymous> (src/testRunner/fourslashRunner.ts:56:39)
      at processImmediate (internal/timers.js:439:21)

```

**Comment by Jack-Works** [comment]
What if I rewrite this pr with the following changes:

- Not rewriting import path anymore. (But still emit files as the designated extension like `.mjs`)
- When `emitExtension` is set, import with no extname is a type error (this will enforce developers to write the target extension like `.mjs`).
- If the extension name !== `emitExtension`, emit a compile error
- Import `.ts` file is still a TypeError (unless the `emitExtension` === ".ts", for the deno case)
- * Add language service fix suggestion to fix wrong extension name when `emitExtension` is set.

About `https://` style import, I have another proposal for this https://github.com/microsoft/TypeScript/issues/35163

@DanielRosenwasser does it nice to ts team?

**Comment by Jack-Works** [comment]
@DanielRosenwasser @RyanCavanaugh Hello, I have fully rewrittened this PR. Please checkout the new behavior! (There is no path rewriting anymore!)

**Comment by MicahZoltu** [comment]
The thing that bugs me about including extensions as enforced by the `noImplicitExtensions` flag is that I don't necessarily know at dev time what the file extensions for my files will be.  I commonly have two `tsconfig.json` files in my project, one for targeting `es2015` modules and one for targeting `commonjs` modules.  If I switch to targeting es2015 modules for both and instead change to having one emit `.js` and the other emit `.mjs` then using something like this will result in one of them generating bad code.  If I hard-code `.js` in my import statements then the tsconfig that emits `.mjs` files will not work.  If I hard-code `.mjs` in my import statements then the tsconfig that emits `.js` files will not work.

**Comment by Jack-Works** [comment]
I'm sorry but the ts team doesn't accept the previous version of this pr that will do the extension name rewrite for you. So I think there is no way to resolve this problem. @MicahZoltu

**Comment by MicahZoltu** [comment]
Yeah, I have been following along and am frustrated with the current situation.  Browser and NodeJS have diverged, and TypeScript is caught in the middle with an unwillingness to resolve that divergence on behalf of users.  This means TypeScript users are forced to "pick a side" and they can no longer write TypeScript that runs in both browser and NodeJS.

My plan for the time being is to continue using a TypeScript transformer to add the extension as part of the TSC build process.  This way I can use a different transformer based on target without having to write TypeScript that has "picked a side".

**Comment by Jack-Works** [comment]
I guess https://github.com/cevek/ttypescript is a good project to start with for your case

**Comment by Jack-Works** [comment]
About deno & browser style import, see: #35749

**Comment by Jack-Works** [comment]
if ts team wants me to rebase my code to the latest codebase, @ me at any time!

**Comment by ExE-Boss** [comment]
> Fixes #30076, #18442, #16577, #33306

[You need to write the "fixes" keyword before each issue to close multiple issues.](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)

**Comment by MrCheater** [comment]
+1

**Comment by sandersn** [comment]
@rbuckton I added you as the main reviewer since #33306 was assigned to you for 3.8.1, 
but @weswigham, can you review this too since you've spent a lot of time thinking about modules?

**Comment by Jack-Works** [comment]
@weswigham @rbuckton 
hi, I have rebased this PR to the latest codebase.
There's something strange happened because the test passed on my Windows and Linux environment but failed on CI.

**Comment by weswigham** [comment]
Our design meeting notes from last Friday will (hopefully) have more details on the discussion we had, but we want to pursue a more complete design for the support of the modern browser resolver, the upcoming node resolver, and the deno resolver. What we wanna do, rather than configuring the current `moduleResolution: "node"` resolver we have, is provide _new_ `moduleResolution` modes for each of these new resolvers. As far as stability goes, I personally think a `moduleResolution: "browser"` which requires exact extensions and does not allow any kind of module lookup is probably stable enough to implement now (and maybe iteratively add some kind of import map support for later). Maybe this could be paired with some rudimentary url support. A `moduleResolution: "node-esm"` (or maybe just `node14`) could be started, but we'll probably wait on merging it at least until export maps stabilize (so export maps and mjs/cjs support don't require separate flags). Along with that resolver mode, we need to consider the story around output extensions and such; we very much want to try to keep the ts -> js mapping, generally speaking, but we also aren't thrilled about a .mts or .cts extension; so these competing priorities make this resolver, specifically, difficult. We could likewise do a `moduleResolution: "deno"` (ideally similar to the browser one, but with different allowed import statement extensions); but what we're unsure of is weather deno's actually popular enough to warrant the work and continued support. (It's obviously neat, but once we add it, we'll be supporting it for a long time - _looks over at systemjs compatible output support_).

**Comment by jkrems** [comment]
> A moduleResolution: "node-esm" (or maybe just node14) could be started, but we'll probably wait on merging it at least until export maps stabilize (so export maps and mjs/cjs support don't require separate flags).

*The following is partially speculation.*

Would that resolution be dependent on the `module` setting? E.g. would it use the `require` condition when targeting `module: "commonjs"`? "Please read the upcoming design meeting notes" is a perfectly acceptable answer of course.

**Comment by weswigham** [comment]
> Would that resolution be dependent on the module setting? E.g. would it use the require condition when targeting module: "commonjs"? 

Probably! What I'm thinking is that `moduleResolution: "node-esm"` (or whatever its name is) should imply `module: es2018` (if it's not set), and then the behavior of source ts files in the compilation should match cjs or esm resolver modes depending on the `module` setting (or, alternatively, depending on the `package.json` `mode` flag and extensions, if we add support for new extensions). Unfortunately that extension question is the major open decision point. :(

**Comment by Jack-Works** [comment]
> As far as stability goes, I personally think a module solution: "browser" which requires exact extensions and does not allow any kind of module lookup is probably stable enough to implement now (and maybe iteratively add some kind of import map support for later). Maybe this could be paired with some rudimentary URL support.

How will the URL import module resolution work? (to reuse the DefinitelyTyped community as possible if the remote URL is a package, or the remote URL is a normal Javascript file (with .d.ts)) Does it have discussed yet? Hope to see that design

**Comment by orta** [comment]
Thanks @Jack-Works 

Just in case someone wants to ever test this build out independently:

@typescript-bot pack this

**Comment by MicahZoltu** [comment]
> As far as stability goes, I personally think a moduleResolution: "browser" **which requires exact extensions** and does not allow any kind of module lookup is probably stable enough to implement now (and maybe iteratively add some kind of import map support for later). Maybe this could be paired with some rudimentary url support.

@weswigham Let me know if there is a better place to continue this discussion rather than this closed issue.

The problem with the above quote (emphasized piece specifically) is that it means someone cannot write a library that is targeted at *both* browser (which requires explicit JS extension) *and* a pure TS runtime (where JS is never emitted) or some future runtime target that isn't JS compile target.  Ideally, a library author can write TypeScript and then have two different `tsconfig.json` files that target different environments.  If `browser` *demands* that all imports have explicit JS extension while some other target *demands* that all imports have no extension, explicit TS extension, or explicit other extension (e.g., `mjs`) then the library *cannot* be cross-compiled for browser and other.

This scenario is frustrating because the compiler has enough information to handle this situation (by doing import rewrites when extension is missing), yet it chooses not to (for reasons I understand, but don't agree are strong enough to overrule the benefit).

I personally run into this problem all the time (and even have a compiler extension to address it) because I target my libraries at browser, but I test them in `ts-node`.  `ts-node` fails if an explicit `.js` extension is provided, while browser fails if no extension is provided.

**Comment by weswigham** [comment]
Yep, deno and browser resolution are essentially mutually exclusive; it's pretty much impossible to write imports that satisfy both runtimes, and we don't want to be in the game of figuring out how to satisfy both simultaneously. (Unlike browser and node-esm, where most browser imports should be compatible with the node resolver, as that's a stated goal of its development).

Multi-output is essentially very out of scope - it has massive checking and program building implications that are nonobvious and costly, and in some cases essentially calls for building some strange custom resolver with unclear rules that doesn't map to any single runtime. Our reccomendation for situations like that as of right now will _probably_ be to use the browser target and output js files in those scenarios; since afaik pretty much all of them still support just loading js files as normal.

**Comment by Jack-Works** [comment]
> it's pretty much impossible to write imports that satisfy both runtimes

No, it's possible and I have done it. I have a [JSON RPC library](https://github.com/Jack-Works/async-call) that has totally 0 npm dependencies and 0 platform dependencies (only use things defined in ECMAScript spec) so I can ship this library for browser, deno, and node.

The only problem is the `import specifier` is host-related (with extension `.ts` or `.js`, or no extension), so I have to write all logic in a single file so there is no `import/export from "./a-local-file"`.

The relative module specifier is somehow shared between browser, node, and deno, besides the extension name. If the TypeScript compiler supports this kind of rewrite, I can write 3 tsconfig for browser, node, and deno, therefore, to generate 3 outputs instead of writing everything in a single file.

-------------------------------------

Even if, there are 3rd party dependencies, it is also possible to write packages for any runtime. Here is [an example](https://github.com/Jack-Works/ttypescript-browser-like-import-transformer/tree/develop/src): In the `src` folder, here is a `core.ts` that has 0 (value) dependencies and exports a factory function to construct the library with 3rd party dependencies.
Here is a `node.ts` to write node-specified code, it imports 3rd party libs in Node style, then imports the `core.ts`, and call the library constructor.

If I need to support deno, I just need to write a `deno.ts` and then import libs in deno style and call the factory.

----------------------------------

Therefore, if TS can support extension name rewriting, things will be easier in the way of writing a lib for every ES environment

**Comment by MicahZoltu** [comment]
> afaik pretty much all of them still support just loading js files as normal.

This is not true.  I currently cannot write code that works in both browser and ts-node.  `ts-node` fails if you include explicit JS extensions (reasonably so IMO) and browser fails if you don't include explicit extensions.

> Multi-output is essentially very out of scope - it has massive checking and program building implications that are nonobvious and costly, and in some cases essentially calls for building some strange custom resolver with unclear rules that doesn't map to any single runtime.

Where can I read more about this?  I have a TypeScript transformer that solves all of the problems I and many others need it to solve and all it does is adds a JS extension to relative import emits.

Perhaps you are trying to solve more problems than is being asked for here?  I think the main thing we want solved is *ONLY* extension emit added to relative imports.  We are not looking for a solution to package imports differences between platforms.  We just want to not have to build our libraries in a single file as @Jack-Works described or use a custom transformer.  At the least, please don't break the custom transformer solution!

**Comment by kitsonk** [comment]
> We could likewise do a moduleResolution: "deno" (ideally similar to the browser one, but with different allowed import statement extensions); but what we're unsure of is weather deno's actually popular enough to warrant the work and continued support.

@weswigham I think we are comfortable with that label (for now 😄).  Deno's resolution is pretty straight forward and there are enough options with the Deno runtime and other plugins that give the resolution mode it is fine.

As stated elsewhere, it is hard to write isomorphic TypeScript code that runs in Deno, and other platforms.  For Deno, we are looking at denoland/deno#4538 to fix situations where you want to publish code for use in other runtimes, when `deno bundle` isn't sufficient.

**Comment by frank-dspeed** [comment]
mjs and cjs support are simply needed no resolver needed it can use the node resolver deno uses the parcel bundler that also supports node resolve.

**Comment by TheMrZZ** [comment]
I don't think I understand correctly all the implications of such a new feature, but its benefits are clear, and massive. I hope we will resolve all problems regarding this, and that it will one day get released!

**Comment by nitsky** [comment]
See also #38149

**Comment by Jack-Works** [comment]
@wemeetagain @orta what about pick the compilerOption `--noImplicitExtensionName` out and make a seperate PR for it? Will ts accept it?

**Comment by orta** [comment]
That one is probably on @weswigham / @RyanCavanaugh - I don't feel like I can reliably say

**Comment by dilyanpalauzov** [comment]
I want to use TypeScript to check the types.  But to compile the code fast I want to strip the types with `@babel/preset-typescript`.

When I write in some of my files `import x from ./x.ts` tsc 3.9.2 says: “An import path cannot end with a .ts extension”, but `rollup`/`@rollup/plugin-babel`/`@babel-preset-typescript` eat the input happy.

When I write instead `import x from ./x.js`, typescript does eat the input, but rollup/`@rollup/plugin-babel`/`@babel/preset-typescript` cannot find the file.  Likewise, when I use `import x from ./x`.

This code, a.js:
```javascript
import b from './b.ts'
console.log(b)
```
b.ts:
```javascript
export default 'blub'
```
is valid JavaScript, it just does not work in Node.JS, as the latter refuses to accept .ts file extension.  

But when b.ts is delivered with the correct mime type, then this works:
```html
<html>
  <body><script type='module' src='./a.js'></script></body>
</html>
```

Demonstration: https://mail.aegee.org/dpa/35148/file.html

To sum up, 
```javascript
import b from './b.ts'
```
is valid JavaScript, and tsc refuses to accept valid JavaScript.

How to get now the same input working in both tsc 3.9.2 and `rollup`/`@rollup/plugin-babel`/`@babel/preset-typescript`?

**Comment by shrinktofit** [comment]
What's the status?

**Comment by Jack-Works** [comment]
Typescript team rejected this idea 😭

**Comment by awkj** [comment]
> 打字稿团队拒绝了这个想法 😭



> Typescript team rejected this idea 😭

Why is typescript so arrogant, self-healing development

**Comment by timreichen** [comment]
Why?

**Comment by knightedcodemonkey** [comment]
There are some options for dealing with this that might suit some use cases described here:

* [@knighted/specifier](https://github.com/knightedcodemonkey/specifier) for changing your ESM/CJS specifiers as you see fit.
* [@knighted/duel](https://github.com/knightedcodemonkey/duel) for creating dual TypeScript packages with correct file and specifier extensions.
* [babel-dual-package](https://github.com/morganney/babel-dual-package) similar to @knighted/duel except using Babel and `@babel/preset-typescript`.
